Gutenboarding: use static, tall screenshots

simison opened the pull request:
Resolves https://github.com/Automattic/wp-calypso/issues/40564

#### Changes proposed in this Pull Request

* Add static screenshots for designs
* Add a script that generated those previews
* Use those screenshots in img tags — leaves mShots mechanism in place so we can easily work on it
* Make screenshots little taller (needs more work)

#### Testing instructions
##### Design
- See [Gutenboarding](https://wpcalypso.wordpress.com/gutenboarding) design picker step:

<img width="1750" alt="image" src="https://user-images.githubusercontent.com/87168/78232734-5ae87680-74dd-11ea-9b3a-761f55af6ac2.png">


##### CLI
- To ensure you'll have browser downloaded, remove `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true` from `package.json`
- At Calypso root, install deps: `npm i capture-website sharp`.
- Remove old screenshots at `static/page-templates/design-screenshots`
- Run `node bin/generate-gutenboarding-design-thumbnails.js` to generate new ones
- Observe new files at `static/page-templates/design-screenshots`

simison committed:
Add screenshots

simison committed:
Add a script to generate screenshots

simison committed:
Use screenshots and make 'em taller

simison committed:
use networkidle0

simison commented on bin/generate-gutenboarding-design-thumbnails.js:
```diff
@@ -0,0 +1,59 @@
+				height: 2200,
+				deviceScaleFactor: 1,
+			} );
+			await page.goto( url, { waitUntil: 'networkidle0' } );
```
Possible options for `waitUntil` would be 

> - `load` - consider navigation to be finished when the load event is fired.
> - `domcontentloaded` - consider navigation to be finished when the DOMContentLoaded event is fired.
> - `networkidle0` - consider navigation to be finished when there are no more than 0 network connections for at least 500 ms.
> - `networkidle2` - consider navigation to be finished when there are no more than 2 network connections for at least 500 ms.

via https://github.com/puppeteer/puppeteer/blob/3dff24ad35cd46754dbe0c77a9c7f7ade2454b88/docs/api.md#pagewaitfornavigationoptions

Commit:
Add breakpoints for design-selector__grid

dubielzyk commented:
Mobile layout looks funky. Let's make sure we look at the design file and replicate the mobile designs:
![image](https://user-images.githubusercontent.com/2896062/78114688-f3112d80-7401-11ea-9b72-408401288b88.png)

Other than that it looks good 👍

simison commented:
We're kinda ignoring the mobile versions for this week and starting their implementation on next week. Ok of course to fix in the PR but ok to punt for now for later PRs.

simison commented on bin/generate-gutenboarding-design-thumbnails.js:
```diff
@@ -0,0 +1,59 @@
+			const page = await browser.newPage();
+			await page.setViewport( {
+				width: 980,
+				height: 2200,
```
Might actually need to set viewport height to something smaller, take the screenshot with `fullPage: false` and then crop it before saving.

alshakero replied:
BTW, I would use [`capture-screen`](https://github.com/sindresorhus/capture-website). Would save you from a lot of headache.

simison replied:
Ah, nice one! I agree but now this script is already done and the viewport settings seem to be just the ones passed to Puppeteer, it won't do cropping for us anyway.

simison replied:
@alshakero thanks for the tip — ended up using that after all.

simison committed:
Update CLI script

- Use `capture-website` instead of Puppeteer directly to simplify
- To crop and resize the output image

simison committed:
Update screenshots

simison committed:
Use flag instead of constant

simison commented:
@dubielzyk Screenshots are sometimes shorter than the thumbnail:

<img width="913" alt="image" src="https://user-images.githubusercontent.com/87168/78233152-009be580-74de-11ea-9b5e-f6c2eb75522d.png">

Luckily they're all the ones with a white background so it's not that bad looking, but do you have thoughts on how to address that? Should columns float as "masonry" (i.e. upper edges won't be aligned and it's ok to have different height thumbnails)?

It should be addressed in separate PR anyway, I think.

dubielzyk commented:
>Screenshots are sometimes shorter than the thumbnail area

I'll have a look, but I think it's a tradeoff worth taking since users wanted to see more of their page

simison commented on bin/generate-gutenboarding-design-thumbnails.js:
```diff
@@ -0,0 +1,80 @@
+ */
+
+const captureWebsite = require( 'capture-website' );
+const designs = require( '../client/landing/gutenboarding/available-designs.json' );
```
Will need a bit of work once https://github.com/Automattic/wp-calypso/pull/40587 merges but not a biggie.

sirreal replied:
Oh… that's going to conflict with some changes in https://github.com/Automattic/wp-calypso/pull/40680 (json -> ts)

simison replied:
This script is so throwaway material that it doesn't matter. We might need to run this just once or twice more (hope not at all) so even just listing designs here manually would be fine.

(designs = URLs like `https://public-api.wordpress.com/rest/v1/template/demo/mayland/mayland2/?font_headings=Cabin&font_base=Raleway`)

sirreal replied:
We concluded via another channel that this can be addressed when the need arises.

simison commented:
Feels like that 4 columns takes affect little too early (somewhere at ~1400px), making thumbnails too small:

<img width="1449" alt="image" src="https://user-images.githubusercontent.com/87168/78236728-13fd7f80-74e3-11ea-9e7e-9afa6a8c4e94.png">

We can get these right in separate PR.

simison commented on bin/generate-gutenboarding-design-thumbnails.js:
```diff
@@ -0,0 +1,80 @@
+
+			// Fix `reynolds_rockfield2_rockfield.jpg` first section becoming super tall
+			// @TODO: fix at the source since this will be an issue with mshots API, too.
+			const styles =
```
Can be completely removed since the template is going away: https://github.com/Automattic/wp-calypso/issues/40700

simison replied:
Nope — I mixed templates. This one is actually not going away. :-)

simison committed:
Remove style override for template that's going to be removed

simison committed:
Revert "Remove style override for template that's going to be removed"

This reverts commit b2954f8c90532ab74d726df8490020cc64881d10.

Commit:
Add fullhd breakpoint and use it for design-selector__grid

sirreal commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -39,13 +39,22 @@ const DesignSelector: React.FunctionComponent = () => {
+		// https://github.com/Automattic/mShots/issues/16
+		// https://github.com/Automattic/wp-calypso/issues/40564
+		if ( ! isEnabled( 'gutenboarding/mshot-preview' ) ) {
+			return `/calypso/page-templates/design-screenshots/${ design.slug }_${ design.template }_${ design.theme }.jpg`;
```
I don't like that we're referencing static images directly rather than using webpack file-loader to import the images and get a generated path. That help with things like cache busting when image contents change.

Are these intended for use outside of Calypso?

This is fine to handle as a follow-up.

sirreal replied:
Created #40708 to track.

razvanpapadopol reviewed (approved):
Looking great on all resolutions! Added #40711 and #40712 as follow-ups so we can 🚀 this.
